File: packages/verification_engine/schemas.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

from packages.intent_compiler.schemas import _reject_string_amount


class ActionType(str, Enum):
    PURCHASE = "purchase"


class ProductRef(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    name: str
    category: str | None = None
    attributes: dict[str, Any] = Field(default_factory=dict)


class LineItem(BaseModel):
    sku: str
    name: str
    amount: Decimal = Field(..., ge=0, decimal_places=2, max_digits=12)
    quantity: int = Field(default=1, ge=1)

    @field_validator("amount", mode="before")
    @classmethod
    def amount_must_be_numeric(cls, value: object) -> object:
        return _reject_string_amount(value)


class ProposedAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    action: ActionType = ActionType.PURCHASE
    amount: Decimal = Field(..., gt=0, decimal_places=2, max_digits=12)
    currency: str = "INR"
    merchant: str
    product: ProductRef
    quantity: int = Field(default=1, ge=1)
    line_items: list[LineItem] = Field(default_factory=list)
    agent_rationale: str | None = None
    scheduled_at: datetime | None = None
    location: str | None = None

    @field_validator("amount", mode="before")
    @classmethod
    def amount_must_be_numeric(cls, value: object) -> object:
        return _reject_string_amount(value)

    @field_serializer("amount")
    def serialize_amount(self, value: Decimal) -> float:
        return float(value)


class SubstitutionSeverity(str, Enum):
    NONE = "none"
    MINOR = "minor"
    MAJOR = "major"


class SemanticAssessment(BaseModel):
    """LLM output only. Never contains a payment decision."""

    model_config = ConfigDict(extra="forbid")

    semantic_match: float | None = Field(default=None, ge=0.0, le=1.0)
    violated_preferences: list[str] = Field(default_factory=list)
    substitution_severity: SubstitutionSeverity = SubstitutionSeverity.NONE
    reason: str = ""
    error: bool = False


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class RiskAssessment(BaseModel):
    risk_level: RiskLevel = RiskLevel.LOW
    injection_high: bool = False
    flags: list[str] = Field(default_factory=list)


class HardConstraintFailure(BaseModel):
    code: str
    message: str


class HardConstraintResult(BaseModel):
    passed: bool
    failures: list[HardConstraintFailure] = Field(default_factory=list)


class Verdict(str, Enum):
    APPROVE = "APPROVE"
    PAUSE = "PAUSE"
    BLOCK = "BLOCK"


class Decision(BaseModel):
    """Written only by the deterministic decision engine."""

    verdict: Verdict
    reasons: list[str] = Field(default_factory=list)
    policy_version: str
    proposal_id: UUID | None = None

File: apps/web/lib/schemas.ts
import { z } from "zod";

const amountSchema = z.number().positive();

export const hardConstraintsSchema = z.object({
  max_amount: amountSchema,
  currency: z.literal("INR"),
  category: z.string().nullable().optional(),
  quantity: z.number().int().positive().default(1),
  allowed_merchants: z.array(z.string()).default([]),
  forbidden_attributes: z.array(z.string()).default([]),
  must_include: z.array(z.string()).default([]),
  time_window: z.string().nullable().optional(),
  location: z.string().nullable().optional(),
});

export const preferencesSchema = z
  .object({
    weight: z.string().nullable().optional(),
    use_case: z.string().nullable().optional(),
    preferred_brands: z.array(z.string()).default([]),
  })
  .passthrough();

export const intentContractSchema = z.object({
  intent_id: z.string().uuid().nullable().optional(),
  created_at: z.string().nullable().optional(),
  raw_request: z.string().min(1),
  goal: z.string().min(1),
  hard_constraints: hardConstraintsSchema,
  preferences: preferencesSchema.default({ preferred_brands: [] }),
  approval_required_for: z.array(z.string()).default([]),
  expires_at: z.string().nullable().optional(),
  status: z.enum([
    "draft",
    "active",
    "paused",
    "approved",
    "blocked",
    "expired",
    "paid",
  ]).default("draft"),
});

export const proposedActionSchema = z.object({
  action: z.literal("purchase").default("purchase"),
  amount: amountSchema,
  currency: z.string().default("INR"),
  merchant: z.string().min(1),
  product: z.object({
    id: z.string(),
    name: z.string(),
    category: z.string().nullable().optional(),
    attributes: z.record(z.string(), z.unknown()).default({}),
  }),
  quantity: z.number().int().positive().default(1),
  line_items: z
    .array(
      z.object({
        sku: z.string(),
        name: z.string(),
        amount: amountSchema,
        quantity: z.number().int().positive().default(1),
      }),
    )
    .default([]),
  agent_rationale: z.string().nullable().optional(),
  scheduled_at: z.string().nullable().optional(),
  location: z.string().nullable().optional(),
});

export const semanticAssessmentSchema = z.object({
  semantic_match: z.number().min(0).max(1).nullable(),
  violated_preferences: z.array(z.string()).default([]),
  substitution_severity: z.enum(["none", "minor", "major"]).default("none"),
  reason: z.string().default(""),
  error: z.boolean().default(false),
});

export const decisionSchema = z.object({
  verdict: z.enum(["APPROVE", "PAUSE", "BLOCK"]),
  reasons: z.array(z.string()).default([]),
  policy_version: z.string(),
  proposal_id: z.string().uuid().nullable().optional(),
});

export type IntentContract = z.infer<typeof intentContractSchema>;
export type ProposedAction = z.infer<typeof proposedActionSchema>;
export type SemanticAssessment = z.infer<typeof semanticAssessmentSchema>;
export type Decision = z.infer<typeof decisionSchema>;
